### PR TITLE
[#4435] Always invoke the actual deregisteration later in the EventLoop.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

As a user may call deregister() from within any method while doing processing in the ChannelPipeline,  we need to ensure we do the actual deregister operation later. This is needed as for example,  we may be in the ByteToMessageDecoder.callDecode(...) method and so still try to do processing in the old EventLoop while the user already registered the Channel to a new EventLoop. Without delay, the deregister operation this could lead to have a handler invoked by different EventLoop and so threads.

Modifications:

Ensure the actual deregister will be done later on and not directly when invoked.

Result:

Calling deregister() within ByteToMessageDecoder.decode(..) is safe.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin @Scottmitch PTAL

**Comment by normanmaurer** [comment]
@trustin @Scottmitch PTAL... Fix was verified by @doom369 who reported the issue before.

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 1]
```diff
@@ -650,27 +645,43 @@ public final void deregister(final ChannelPromise promise) {
...
+                        // Some transports like local and AIO does not allow the deregistration of
+                        // an open channel.  Their doDeregister() calls close(). Consequently,
+                        // close() calls deregister() again - no need to fire channelUnregistered, so check
+                        // if it was registered.
+                        if (registered) {
+                            registered = false;
+                            invokeLater(new OneTimeTask() {
+                                @Override
+                                public void run() {
+                                    pipeline.fireChannelUnregistered();
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: We are already in `OneTimeTask`. Perhaps we could call `pipeline.fireChannelUnregistered()` directly?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: @trustin I thought it was "delayed" for some reason in the past but could not really remember, so I thought I should just  keep it delayed. But if you think it is fine to directly call the method after pipeline.fireChannelInactive() I'm happy to do so. WDYT ?

Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: We are already in invokeLater, so I don't think we need to do it again. i.e. direct call should be OK IMO.

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: alright... let me update the PR and see if we get any issues :)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (15162202fb82e2293624a86bfc27a9c5c35960be) and 4.1 (d59bf84ef2670ab830782f8e8036f0b3c1f46f34)

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: +1

**Comment by Scottmitch** [comment]
lgtm ... good find!

**Pull request closed without merging**